File: 283. Move Zeroes/one_traversal.cpp
/*
Time complexity: O(n)
Space complexity: O(1)
*/
class Solution {
public:
    void moveZeroes(vector<int>& nums, int val = 0) {
        for (int lastNonZeroFoundAt = 0, cur = 0; cur < nums.size(); cur++) {
            if (nums[cur] != val) { // если заменить на "==", то val перенесется в префикс
                swap(nums[lastNonZeroFoundAt++], nums[cur]);
            }
        }
    }
};

File: 283. Move Zeroes/one_traversal.go
func moveZeroes(nums []int) {
	Val := 0 
	PtrOnSwapFirstVal := 0 // == Count(NotVal)
    // в любом префиксе действует инвариант: Count(Val) + Count(NotVal) == i
    // при Count(Val) != 0 между PtrOnSwapFirstVal и i возникает окно из Val
    // длиною Count(Val), при это PtrOnSwapFirstVal всегда указывет на первый
    // элемент этого окна
	for i := 0; i < len(nums); i++ {
		if nums[i] != Val { // // != -->  [1,2,3,4]
			nums[i], nums[PtrOnSwapFirstVal] = nums[PtrOnSwapFirstVal], nums[i]
            PtrOnSwapFirstVal++
			
		}
	}
}
